### PR TITLE
Add delegating ticket store

### PR DESCRIPTION
**Pull request by pradtke**
Allow delegating to multiple other ticket stores to enable
a data store transition without down time or to do increase
redundancy.

### CODE REVIEW

File: lib/Cas/Ticket/DelegatingTicketStore.php [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+use SimpleSAML\Logger;
+use SimpleSAML\Module;
+
+/**
+ * Ticket store that delegates to other ticket stores.
+ * Allows you to use multiple source for redundancy or allows a transition from one source to another.
+ */
+class DelegatingTicketStore extends TicketStore
+{
+    private $delegateTo = 'all';
```
Comment (tvdijen) [inline lib/Cas/Ticket/DelegatingTicketStore.php]: /** @var string */

File: lib/Cas/Ticket/DelegatingTicketStore.php [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+     * @var TicketStore[]
+     */
+    private $ticketStores = [];
+
+    /**
+     * @var TicketStore
+     */
+    private $primaryDelegate;
+
+    public function __construct(Configuration $casConfig)
```
Comment (tvdijen) [inline lib/Cas/Ticket/DelegatingTicketStore.php]: /** @param \SimpleSAML\Configuration $casConfig */

File: lib/Cas/Ticket/DelegatingTicketStore.php [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                throw $rethrowException;
+            }
+            return $ticket;
+        } else {
+            return $this->primaryDelegate->getTicket($ticketId);
+        }
+    }
+
+    /**
+     * @param $ticket
```
Comment (tvdijen) [inline lib/Cas/Ticket/DelegatingTicketStore.php]: @param array $ticket

File: lib/Cas/Ticket/DelegatingTicketStore.php [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+                $this->primaryDelegate = $ticketStores[$this->delegateTo];
+            } else {
+                throw new InvalidArgumentException("No ticket store called '" . $this->delegateTo . "'");
+            }
+        }
+    }
+
+    /**
+     * Get the ticket, searching one or all of the delegates
+     * @param $ticketId string The ticket to find
```
Comment (tvdijen) [inline lib/Cas/Ticket/DelegatingTicketStore.php]: type goes before variable name

File: lib/Cas/Ticket/DelegatingTicketStore.php [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            if ($rethrowException) {
+                throw $rethrowException;
+            }
+        } else {
+            $this->primaryDelegate->addTicket($ticket);
+        }
+    }
+
+    /**
+     * @param $ticketId string
```
Comment (tvdijen) [inline lib/Cas/Ticket/DelegatingTicketStore.php]: type goes before variable

File: tests/lib/Ticket/DelegatingTicketStoreTest.php [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+     * @var TicketStore $fileStore1
+     */
+    private $fileStore1;
+
+    /**
+     * @var TicketStore $fileStore2
+     */
+    private $fileStore2;
+
+    public function setup(): void
```
Comment (tvdijen) [inline tests/lib/Ticket/DelegatingTicketStoreTest.php]: @return void (for all tests in this file)

File: tests/lib/Ticket/ErroringTicketStore.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+namespace Simplesamlphp\Casserver\Ticket;
+
+use SimpleSAML\Module\casserver\Cas\Ticket\TicketStore;
+
+/**
+ * Ticket store that always generates errors, used for testing error handling
+ */
+class ErroringTicketStore extends TicketStore
+{
+
```
Comment (tvdijen) [inline tests/lib/Ticket/ErroringTicketStore.php]: Excess whitespace (PSR-12)

File: lib/Cas/Ticket/DelegatingTicketStore.php [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    private $primaryDelegate;
+
+    /**
+     * @param \SimpleSAML\Configuration $casConfig The cas configuration.
+     */
+    public function __construct(Configuration $casConfig)
+    {
+        $config = $casConfig->getConfigItem('ticketstore');
+        $this->delegateTo = $config->getString('delegateTo', 'all');
+        /** @var $storeConfig Configuration */
```
Comment (tvdijen) [inline lib/Cas/Ticket/DelegatingTicketStore.php]: This annotation is wrong as well and I don't think it's necessary at all

**Comment by pradtke** [comment]
Thanks @tvdijen for the review!

**Pull request merged**